microcosm_pubsub/result.py

```diff
@@ -42,69 +44,102 @@ class MessageHandlingResultType(Enum):
     def __str__(self):
         return self.name
 
+    @property
+    def exc_info(self):
+        return self.value.exc_info
+
     @property
     def level(self):
         return self.value.level
 
     @property
-    def exc_info(self):
-        return self.value.exc_info
+    def retry(self):
+        return self.value.retry
 
 
 @dataclass
 class MessageHandlingResult:
-    exc_info: Tuple[Any, Any, Any]
-    extra: Dict[str, str]
     media_type: str
     result: MessageHandlingResultType
-    elapsed_time: float = None
+    exc_info: Optional[Tuple[Any, Any, Any]] = None
+    extra: Dict[str, str] = field(default_factory=dict)
+    elapsed_time: Optional[float] = None
+    retry_timeout_seconds: Optional[int] = None
 
     @classmethod
-    def invoke(cls, func, message, **kwargs):
-        exc_info_ = None
-        extra = dict()
+    def invoke(cls, handler, message: SQSMessage):
         try:
-            success = func(message=message, **kwargs)
-            if success:
-                result = MessageHandlingResultType.SUCCEEDED
+            success = handler(message.content)
+            result = cls.from_result(message, bool(success))
```

Comment:
I don't know if I like the explicit casting (I know - hints - but still)

Reply:
I agree, but is it worse to:

Reply:
b

Reply:
You can also declare `handler` as `Callable[[dict], bool]` ([doc](https://docs.python.org/3/library/typing.html#typing.Callable)), and `success: bool = [...]`. Then you shouldn't need to cast it explicitly.

Reply:
But the handler contract isn't actually meant to return `bool`, just something truthy

Reply:
Really? Most of the handlers I had seen returned True/False explicitly.